### ISSUE TITLE
grpc Channel called pure virtual method when program exit

### ISSUE THREAD

**Issue by velconia** [grpc/grpc#15973]
Here is my system info:
```
LSB Version:    :core-4.0-amd64:core-4.0-noarch:graphics-4.0-amd64:graphics-4.0-noarch:printing-4.0-amd64:printing-4.0-noarch
Distributor ID: CentOS
Description:    CentOS release 6.3 (Final)
Release:        6.3
Codename:       Final
```

Here is my gcc building environment:
```
COLLECT_GCC=gcc
Target: x86_64-baidu-linux-gnu
Configured with: ../configure CFLAGS=-O2 --host=x86_64-baidu-linux-gnu --target=x86_64-baidu-linux-gnu --prefix=/top --with-sysroot --with-local-prefix=/ --with-native-system-header-dir=/top/include --with-build-sysroot=/top --with-system-libunwind --enable-languages=c,c++ --enable-c99 --enable-threads=posix --enable-__cxa_atexit --enable-libstdcxx-allocator --enable-libstdcxx-threads --enable-libstdcxx-time --enable-libstdcxx-visibility --disable-multilib --disable-bootstrap
Thread model: posix
gcc version 4.8.2 (GCC) 
```

When my C++ program, which invokes ```grpc``` client, called via a pybind11, everything works fine, but when my program exit, here is the stack:
```
pure virtual method called
terminate called without an active exception
*** Aborted at 1531226147 (unix time) try "date -d @1531226147" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x1f40000b2c5) received by PID 45765 (TID 0x7f8c394c3740) from PID 45765; stack trace: ***
    @     0x7f8c38cd3130 (unknown)
    @     0x7f8c3822f9d9 __GI_raise
    @     0x7f8c382310e8 __GI_abort
    @     0x7f8c0dad3b39 __gnu_cxx::__verbose_terminate_handler()
    @     0x7f8c0dad21fb __cxxabiv1::__terminate()
    @     0x7f8c0dad2234 std::terminate()
    @     0x7f8c0dad2c05 __cxa_pure_virtual
    @     0x7f8bf7567626 grpc::Channel::~Channel()
    @     0x7f8bf75676d9 grpc::Channel::~Channel()
    @     0x7f8bf67955e9 std::_Sp_counted_base<>::_M_release()
    @     0x7f8bf7529f8e distributed::GRPCClient::~GRPCClient()
    @     0x7f8bf752a0a1 distributed::GRPCClient::~GRPCClient()
    @     0x7f8bf7531b9a std::unique_ptr<>::~unique_ptr()
    @     0x7f8c38233269 __run_exit_handlers
    @     0x7f8c382332b5 __GI_exit
    @     0x7f8c38fdee61 Py_Exit
    @     0x7f8c38fdef73 handle_system_exit.part.2
    @     0x7f8c38fdf1e5 PyErr_PrintEx
    @     0x7f8c38fdffb2 PyRun_SimpleFileExFlags
    @     0x7f8c38ff1d3c Py_Main
    @     0x7f8c3821bb45 __libc_start_main
    @     0x7f8c394e087f (unknown)
(END) 

```

After debug for a long time, I found that grpc::Channel inherits from grpc::GrpcLibraryCodegen, and this class uses a global variable which will be destroyed before Channel;

Hope for the fix up.

**Comment by velconia** [comment]
I submit a PR for a hot fix, which will fix this problem quickly.

**Comment by mehrdada** [comment]
@dgquintas this sounds like a C++ issue, and Python usage is only incidental

**Comment by dgquintas** [comment]
If the fix in #15974 is any indication, you may want to try using a more recent version. @nicolasnoble also gave some guidance on that PR.
